<launch>
  <arg name="model" default="$(env TURTLEBOT3_MODEL)" doc="model type [burger, waffle, waffle_pi]"/>
  <arg name="first_tb3"  default="tb3_0"/>
  <arg name="second_tb3" default="tb3_1"/>
  <arg name="third_tb3"  default="tb3_2"/>
  <arg name="fourth_tb3"  default="tb3_3"/>
  <arg name="fifth_tb3"  default="tb3_4"/>
  <arg name="sixth_tb3"  default="tb3_5"/>
  <arg name="seventh_tb3"  default="tb3_6"/>
  <arg name="eighth_tb3"  default="tb3_7"/>
  <arg name="nineth_tb3"  default="tb3_8"/>
  <arg name="tenth_tb3"  default="tb3_9"/>
  <arg name="eleventh_tb3"  default="tb3_10"/>
  <arg name="twelfth_tb3"  default="tb3_11"/>

   <arg name="first_tb3_x_pos" default=" 7.0"/>
  <arg name="first_tb3_y_pos" default=" 6.5"/>
  <arg name="first_tb3_z_pos" default=" 0.0"/>
  <arg name="first_tb3_yaw"   default=" 0.0"/>

  <arg name="second_tb3_x_pos" default=" 8.0"/>
  <arg name="second_tb3_y_pos" default=" 2.0"/>
  <arg name="second_tb3_z_pos" default=" 0.0"/>
  <arg name="second_tb3_yaw"   default=" 0.0"/>

  <arg name="third_tb3_x_pos" default=" 5.5"/>
  <arg name="third_tb3_y_pos" default=" 4.5"/>
  <arg name="third_tb3_z_pos" default=" 0.0"/>
  <arg name="third_tb3_yaw"   default=" 0.0"/>

  <arg name="fourth_tb3_x_pos" default=" -7.0"/>
  <arg name="fourth_tb3_y_pos" default=" 5.0"/>
  <arg name="fourth_tb3_z_pos" default=" 0.0"/>
  <arg name="fourth_tb3_yaw"   default=" 0.0"/>

  <arg name="fifth_tb3_x_pos" default=" 3.0"/>
  <arg name="fifth_tb3_y_pos" default=" -9.0"/>
  <arg name="fifth_tb3_z_pos" default=" 0.0"/>
  <arg name="fifth_tb3_yaw"   default=" 0.0"/>

  <arg name="sixth_tb3_x_pos" default=" -5.0"/>
  <arg name="sixth_tb3_y_pos" default=" -8.0"/>
  <arg name="sixth_tb3_z_pos" default=" 0.0"/>
  <arg name="sixth_tb3_yaw"   default=" 0.0"/>

  <arg name="seventh_tb3_x_pos" default=" -11.0"/>
  <arg name="seventh_tb3_y_pos" default=" -8.0"/>
  <arg name="seventh_tb3_z_pos" default=" 0.0"/>
  <arg name="seventh_tb3_yaw"   default=" 0.0"/>

  <arg name="eighth_tb3_x_pos" default=" -5.0"/>
  <arg name="eighth_tb3_y_pos" default=" -15.0"/>
  <arg name="eighth_tb3_z_pos" default=" 0.0"/>
  <arg name="eighth_tb3_yaw"   default=" 0.0"/>

  <arg name="nineth_tb3_x_pos" default=" 9.0"/>
  <arg name="nineth_tb3_y_pos" default=" 14.0"/>
  <arg name="nineth_tb3_z_pos" default=" 0.0"/>
  <arg name="nineth_tb3_yaw"   default=" 0.0"/>

  <arg name="tenth_tb3_x_pos" default=" 6.0"/>
  <arg name="tenth_tb3_y_pos" default=" 13.0"/>
  <arg name="tenth_tb3_z_pos" default=" 0.0"/>
  <arg name="tenth_tb3_yaw"   default=" 0.0"/>

  <arg name="eleventh_tb3_x_pos" default=" 16.0"/>
  <arg name="eleventh_tb3_y_pos" default=" -8.0"/>
  <arg name="eleventh_tb3_z_pos" default=" 0.0"/>
  <arg name="eleventh_tb3_yaw"   default=" 0.0"/>

  <arg name="twelfth_tb3_x_pos" default=" -12.0"/>
  <arg name="twelfth_tb3_y_pos" default=" 18.0"/>
  <arg name="twelfth_tb3_z_pos" default=" 0.0"/>
  <arg name="twelfth_tb3_yaw"   default=" 0.0"/>

  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <!-- <arg name="world_name" value="$(find turtlebot3_gazebo)/worlds/turtlebot3_house.world"/> -->
    <arg name="paused" value="false"/>
    <arg name="use_sim_time" value="true"/>
    <arg name="gui" value="true"/>
    <arg name="headless" value="false"/>
    <arg name="debug" value="false"/>
  </include>  

  <group ns = "$(arg first_tb3)">
    <param name="robot_description" command="$(find xacro)/xacro --inorder $(find turtlebot3_description)/urdf/turtlebot3_$(arg model).urdf.xacro" />

    <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" output="screen">
      <param name="publish_frequency" type="double" value="50.0" />
      <param name="tf_prefix" value="$(arg first_tb3)" />
    </node>
    
    <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-urdf -model $(arg first_tb3) -x $(arg first_tb3_x_pos) -y $(arg first_tb3_y_pos) -z $(arg first_tb3_z_pos) -Y $(arg first_tb3_yaw) -param robot_description" />
    <node pkg="swarm_aggregation" type="dynamic_with_CA.py" name="multi_bot1" output="screen">
      <remap from="/odom" to="/$(arg first_tb3)/odom"/>
      <remap from="/cmd_vel" to="/$(arg first_tb3)/cmd_vel"/>
      <remap from="/goal" to="/$(arg first_tb3)/goal"/>
    </node>
  </group>


  <group ns = "$(arg second_tb3)">
    <param name="robot_description" command="$(find xacro)/xacro --inorder $(find turtlebot3_description)/urdf/turtlebot3_$(arg model).urdf.xacro" />

    <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" >
      <param name="publish_frequency" type="double" value="50.0" />
      <param name="tf_prefix" value="$(arg second_tb3)" />
    </node>

    <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-urdf -model $(arg second_tb3) -x $(arg second_tb3_x_pos) -y $(arg second_tb3_y_pos) -z $(arg second_tb3_z_pos) -Y $(arg second_tb3_yaw) -param robot_description" />
    <node pkg="swarm_aggregation" type="dynamic_with_CA.py" name="multi_bot2" >
      <remap from="/odom" to="/$(arg second_tb3)/odom"/>
      <remap from="/cmd_vel" to="/$(arg second_tb3)/cmd_vel"/>
      <remap from="/goal" to="/$(arg second_tb3)/goal"/>
    </node>
  </group>

  <group ns = "$(arg third_tb3)">
    <param name="robot_description" command="$(find xacro)/xacro --inorder $(find turtlebot3_description)/urdf/turtlebot3_$(arg model).urdf.xacro" />

    <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" >
      <param name="publish_frequency" type="double" value="50.0" />
      <param name="tf_prefix" value="$(arg third_tb3)" />
    </node>

    <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-urdf -model $(arg third_tb3) -x $(arg third_tb3_x_pos) -y $(arg third_tb3_y_pos) -z $(arg third_tb3_z_pos) -Y $(arg third_tb3_yaw) -param robot_description" />
    <node pkg="swarm_aggregation" type="dynamic_with_CA.py" name="multi_bot3" >
      <remap from="/odom" to="/$(arg third_tb3)/odom"/>
      <remap from="/cmd_vel" to="/$(arg third_tb3)/cmd_vel"/>
      <remap from="/goal" to="/$(arg third_tb3)/goal"/>
    </node>
  </group>

  <group ns = "$(arg fourth_tb3)">
    <param name="robot_description" command="$(find xacro)/xacro --inorder $(find turtlebot3_description)/urdf/turtlebot3_$(arg model).urdf.xacro" />

    <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" output="screen">
      <param name="publish_frequency" type="double" value="50.0" />
      <param name="tf_prefix" value="$(arg fourth_tb3)" />
    </node>
    
    <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-urdf -model $(arg fourth_tb3) -x $(arg fourth_tb3_x_pos) -y $(arg fourth_tb3_y_pos) -z $(arg fourth_tb3_z_pos) -Y $(arg fourth_tb3_yaw) -param robot_description" />
    <node pkg="swarm_aggregation" type="dynamic_with_CA.py" name="multi_bot4" output="screen">
      <remap from="/odom" to="/$(arg fourth_tb3)/odom"/>
      <remap from="/cmd_vel" to="/$(arg fourth_tb3)/cmd_vel"/>
      <remap from="/goal" to="/$(arg fourth_tb3)/goal"/>
    </node>
  </group>

  <group ns = "$(arg fifth_tb3)">
    <param name="robot_description" command="$(find xacro)/xacro --inorder $(find turtlebot3_description)/urdf/turtlebot3_$(arg model).urdf.xacro" />

    <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" output="screen">
      <param name="publish_frequency" type="double" value="50.0" />
      <param name="tf_prefix" value="$(arg fifth_tb3)" />
    </node>
    
    <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-urdf -model $(arg fifth_tb3) -x $(arg fifth_tb3_x_pos) -y $(arg fifth_tb3_y_pos) -z $(arg fifth_tb3_z_pos) -Y $(arg fifth_tb3_yaw) -param robot_description" />
    <node pkg="swarm_aggregation" type="dynamic_with_CA.py" name="multi_bot5" output="screen">
      <remap from="/odom" to="/$(arg fifth_tb3)/odom"/>
      <remap from="/cmd_vel" to="/$(arg fifth_tb3)/cmd_vel"/>
      <remap from="/goal" to="/$(arg fifth_tb3)/goal"/>
    </node>
  </group>

  <group ns = "$(arg sixth_tb3)">
    <param name="robot_description" command="$(find xacro)/xacro --inorder $(find turtlebot3_description)/urdf/turtlebot3_$(arg model).urdf.xacro" />

    <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" >
      <param name="publish_frequency" type="double" value="50.0" />
      <param name="tf_prefix" value="$(arg sixth_tb3)" />
    </node>
    
    <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-urdf -model $(arg sixth_tb3) -x $(arg sixth_tb3_x_pos) -y $(arg sixth_tb3_y_pos) -z $(arg sixth_tb3_z_pos) -Y $(arg sixth_tb3_yaw) -param robot_description" />
    <node pkg="swarm_aggregation" type="dynamic_with_CA.py" name="multi_bot6" >
      <remap from="/odom" to="/$(arg sixth_tb3)/odom"/>
      <remap from="/cmd_vel" to="/$(arg sixth_tb3)/cmd_vel"/>
      <remap from="/goal" to="/$(arg sixth_tb3)/goal"/>
    </node>
  </group> 

  <group ns = "$(arg seventh_tb3)">
    <param name="robot_description" command="$(find xacro)/xacro --inorder $(find turtlebot3_description)/urdf/turtlebot3_$(arg model).urdf.xacro" />

    <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" >
      <param name="publish_frequency" type="double" value="50.0" />
      <param name="tf_prefix" value="$(arg seventh_tb3)" />
    </node>
    
    <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-urdf -model $(arg seventh_tb3) -x $(arg seventh_tb3_x_pos) -y $(arg seventh_tb3_y_pos) -z $(arg seventh_tb3_z_pos) -Y $(arg seventh_tb3_yaw) -param robot_description" />
    <node pkg="swarm_aggregation" type="dynamic_with_CA.py" name="multi_bot6" >
      <remap from="/odom" to="/$(arg seventh_tb3)/odom"/>
      <remap from="/cmd_vel" to="/$(arg seventh_tb3)/cmd_vel"/>
      <remap from="/goal" to="/$(arg seventh_tb3)/goal"/>
    </node>
  </group> 

  <group ns = "$(arg eighth_tb3)">
    <param name="robot_description" command="$(find xacro)/xacro --inorder $(find turtlebot3_description)/urdf/turtlebot3_$(arg model).urdf.xacro" />

    <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" >
      <param name="publish_frequency" type="double" value="50.0" />
      <param name="tf_prefix" value="$(arg eighth_tb3)" />
    </node>
    
    <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-urdf -model $(arg eighth_tb3) -x $(arg eighth_tb3_x_pos) -y $(arg eighth_tb3_y_pos) -z $(arg eighth_tb3_z_pos) -Y $(arg eighth_tb3_yaw) -param robot_description" />
    <node pkg="swarm_aggregation" type="dynamic_with_CA.py" name="multi_bot6" >
      <remap from="/odom" to="/$(arg eighth_tb3)/odom"/>
      <remap from="/cmd_vel" to="/$(arg eighth_tb3)/cmd_vel"/>
      <remap from="/goal" to="/$(arg eighth_tb3)/goal"/>
    </node>
  </group> 

  <group ns = "$(arg nineth_tb3)">
    <param name="robot_description" command="$(find xacro)/xacro --inorder $(find turtlebot3_description)/urdf/turtlebot3_$(arg model).urdf.xacro" />

    <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" >
      <param name="publish_frequency" type="double" value="50.0" />
      <param name="tf_prefix" value="$(arg nineth_tb3)" />
    </node>
    
    <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-urdf -model $(arg nineth_tb3) -x $(arg nineth_tb3_x_pos) -y $(arg nineth_tb3_y_pos) -z $(arg nineth_tb3_z_pos) -Y $(arg nineth_tb3_yaw) -param robot_description" />
    <node pkg="swarm_aggregation" type="dynamic_with_CA.py" name="multi_bot6" >
      <remap from="/odom" to="/$(arg nineth_tb3)/odom"/>
      <remap from="/cmd_vel" to="/$(arg nineth_tb3)/cmd_vel"/>
      <remap from="/goal" to="/$(arg nineth_tb3)/goal"/>
    </node>
  </group>

  <group ns = "$(arg tenth_tb3)">
    <param name="robot_description" command="$(find xacro)/xacro --inorder $(find turtlebot3_description)/urdf/turtlebot3_$(arg model).urdf.xacro" />

    <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" >
      <param name="publish_frequency" type="double" value="50.0" />
      <param name="tf_prefix" value="$(arg tenth_tb3)" />
    </node>
    
    <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-urdf -model $(arg tenth_tb3) -x $(arg tenth_tb3_x_pos) -y $(arg tenth_tb3_y_pos) -z $(arg tenth_tb3_z_pos) -Y $(arg tenth_tb3_yaw) -param robot_description" />
    <node pkg="swarm_aggregation" type="dynamic_with_CA.py" name="multi_bot6" >
      <remap from="/odom" to="/$(arg tenth_tb3)/odom"/>
      <remap from="/cmd_vel" to="/$(arg tenth_tb3)/cmd_vel"/>
      <remap from="/goal" to="/$(arg tenth_tb3)/goal"/>
    </node>
  </group>

  <group ns = "$(arg eleventh_tb3)">
    <param name="robot_description" command="$(find xacro)/xacro --inorder $(find turtlebot3_description)/urdf/turtlebot3_$(arg model).urdf.xacro" />

    <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" >
      <param name="publish_frequency" type="double" value="50.0" />
      <param name="tf_prefix" value="$(arg eleventh_tb3)" />
    </node>
    
    <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-urdf -model $(arg eleventh_tb3) -x $(arg eleventh_tb3_x_pos) -y $(arg eleventh_tb3_y_pos) -z $(arg eleventh_tb3_z_pos) -Y $(arg eleventh_tb3_yaw) -param robot_description" />
    <node pkg="swarm_aggregation" type="dynamic_with_CA.py" name="multi_bot6" >
      <remap from="/odom" to="/$(arg eleventh_tb3)/odom"/>
      <remap from="/cmd_vel" to="/$(arg eleventh_tb3)/cmd_vel"/>
      <remap from="/goal" to="/$(arg eleventh_tb3)/goal"/>
    </node>
  </group>

  <group ns = "$(arg twelfth_tb3)">
    <param name="robot_description" command="$(find xacro)/xacro --inorder $(find turtlebot3_description)/urdf/turtlebot3_$(arg model).urdf.xacro" />

    <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" >
      <param name="publish_frequency" type="double" value="50.0" />
      <param name="tf_prefix" value="$(arg twelfth_tb3)" />
    </node>
    
    <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-urdf -model $(arg twelfth_tb3) -x $(arg twelfth_tb3_x_pos) -y $(arg twelfth_tb3_y_pos) -z $(arg twelfth_tb3_z_pos) -Y $(arg twelfth_tb3_yaw) -param robot_description" />
    <node pkg="swarm_aggregation" type="dynamic_with_CA.py" name="multi_bot6" >
      <remap from="/odom" to="/$(arg twelfth_tb3)/odom"/>
      <remap from="/cmd_vel" to="/$(arg twelfth_tb3)/cmd_vel"/>
      <remap from="/goal" to="/$(arg twelfth_tb3)/goal"/>
    </node>
  </group>

  <node pkg="swarm_aggregation" type="detect_bot.py" name="obs_data"/>
  <node pkg="swarm_aggregation" type="plott.py" name="animator"/>

</launch>
